storage/spanstore/decorator/rate_limit.go

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package decorator
+
+import (
+	"time"
+
+	"github.com/jaegertracing/jaeger/internal/utils"
+	"github.com/jaegertracing/jaeger/model"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+)
+
+// NewRateLimitedWriter decorates a Writer with a rate limiter in order to limit
+// the number of writes per second.
+func NewRateLimitedWriter(writer spanstore.Writer, opts ...Option) (spanstore.Writer, error) {
+	o := applyOptions(opts...)
+	rateLimiter, err := utils.NewRateLimiter(o.writesPerSecond, o.maxBurst)
+	if err != nil {
+		return nil, err
+	}
+	return &rateLimitedWriter{
+		w: writer,
+		l: rateLimiter,
+	}, nil
+}
+
+type options struct {
+	writesPerSecond float64
+	maxBurst        float64
+}
+
+// Option declares a type used to configure the rate limited writer.
+type Option func(*options)
+
+// WritesPerSecond configures the maximum number of writes per second.
+func WritesPerSecond(writesPerSecond float64) Option {
+	return func(o *options) {
+		o.writesPerSecond = writesPerSecond
+	}
+}
+
+// MaxBurst configures the maximum writes than can occur in one burst (i.e.
+// maximum number of tokens that can accrue in token bucket).
+func MaxBurst(maxBurst float64) Option {
+	return func(o *options) {
+		o.maxBurst = maxBurst
+	}
+}
+
+func applyOptions(opts ...Option) options {
+	var o options
+	for _, option := range opts {
+		option(&o)
+	}
+	return o
+}
+
+type rateLimitedWriter struct {
+	w spanstore.Writer
+	l utils.RateLimiter
+}
+
+// WriteSpan wraps the write span method of the inner writer, but invokes the
+// rate limiter before each write.
+func (r *rateLimitedWriter) WriteSpan(s *model.Span) error {
+	const cost = 1
+	for ok, waitTime := r.l.CheckCredit(cost); !ok; ok, waitTime = r.l.CheckCredit(cost) {
```

Comment:
Won't 
```
for ok, waitTime := r.l.CheckCredit(cost); !ok {
   time.Sleep(waitTime)
```
work? Not sure why the second call is needed

Reply:
Well let's imagine Java/C++/wtvr other language:

```
for (bool ok = r.l.CheckCredit(cost); !ok; ok = r.l.CheckCredit(cost)) {
    // ...
}
```

First statement is a declaration/initialization statement. Last statement repeats each time the loop is repeated. Not 100% sure, but I imagine Go works the same way.

Reply:
Ya, this explains it: https://tour.golang.org/flowcontrol/1.